File: Sporthor/Sporthor/Services/SocialService.swift
//
//  SocialService.swift
//  Sporthor
//
//  Created by derTurke on 12.04.2025.
//

import NetworkKit

enum SocialService {
    case search(_ request: [String: Any])
    case addSearchHistory(_ request: [String: Any])
    case getSearchHistory
    case removeSearchHistory(_ request: [String: Any])
    case createPost(_ request: [String: Any])
    case getFeedAsync(_ request: [String: Any])
    case likePost(_ request: [String: Any])
    case unlikePost(_ request: [String: Any])
    case getStoryFeed
    case createStory(_ request: [String: Any])
    case getComments(_ request: [String: Any])
    case addComment(_ request: [String: Any])
    case watchedStory(_ request: [String: Any])
    case hidePost(_ request: [String: Any])
    case deletePost(_ request: [String: Any])
    case reportPost(_ request: [String: Any])
    case deleteStory(_ request: [String: Any])
    case getUserPostsAsync(_ request: [String: Any])
    case confirmationFollow(_ request: [String: Any])
    case getBlockUser
    case addBlockUser(_ request: [String: Any])
    case removeBlockUser(_ request: [String: Any])
}

extension SocialService: NetworkService {
    var path: String {
        switch self {
        case .search:
            return "/api/Social/Search"
        case .addSearchHistory:
            return "/api/Social/AddSearchHistory"
        case .getSearchHistory:
            return "/api/Social/GetSearchHistory"
        case .removeSearchHistory:
            return "/api/Social/RemoveSearchHistory"
        case .createPost:
            return "/api/Social/CreatePost"
        case .getFeedAsync:
            return "/api/Social/GetFeedAsync"
        case .likePost:
            return "/api/Social/LikePost"
        case .unlikePost:
            return "/api/Social/UnlikePost"
        case .getStoryFeed:
            return "/api/Social/GetStoryFeed"
        case .createStory:
            return "/api/Social/CreateStory"
        case .getComments:
            return "/api/Social/GetComments"
        case .addComment:
            return "/api/Social/AddComment"
        case .watchedStory:
            return "/api/Social/WatchedStory"
        case .hidePost:
            return "/api/Social/HidePost"
        case .deletePost:
            return "/api/Social/DeletePost"
        case .reportPost:
            return "/api/Social/ReportPost"
        case .deleteStory:
            return "/api/Social/DeleteStory"
        case .getUserPostsAsync:
            return "/api/Social/GetUserPostsAsync"
        case .confirmationFollow:
            return "/api/Social/ConfirmationFollow"
        case .getBlockUser:
            return "/api/Social/GetBlockUser"
        case .addBlockUser:
            return "/api/Social/AddBlockUser"
        case .removeBlockUser:
            return "/api/Social/RemoveBlockUser"
        }
    }
    
    var method: NetworkKit.HTTPMethod {
        switch self {
        case .addSearchHistory,
                .removeSearchHistory,
                .createPost,
                .likePost,
                .unlikePost,
                .createStory,
                .addComment,
                .watchedStory,
                .hidePost,
                .deletePost,
                .reportPost,
                .deleteStory,
                .confirmationFollow,
                .addBlockUser,
                .removeBlockUser:
            return .POST
        default:
            return .GET
        }
    }
    
    var parameters: [String : Any]? {
        switch self {
        case .search(let request):
            return request
        case .addSearchHistory(let request):
            return request
        case .removeSearchHistory(let request):
            return request
        case .createPost(let request):
            return request
        case .getFeedAsync(let request):
            return request
        case .likePost(let request):
            return request
        case .unlikePost(let request):
            return request
        case .createStory(let request):
            return request
        case .getComments(let request):
            return request
        case .addComment(let request):
            return request
        case .watchedStory(let request):
            return request
        case .hidePost(let request):
            return request
        case .deletePost(let request):
            return request
        case .reportPost(let request):
            return request
        case .deleteStory(let request):
            return request
        case .getUserPostsAsync(let request):
            return request
        case .confirmationFollow(let request):
            return request
        case .addBlockUser(let request):
            return request
        case .removeBlockUser(let request):
            return request
        default:
            return nil
        }
    }
    
    var token: String? {
        switch self {
        case .addSearchHistory,
                .getSearchHistory,
                .removeSearchHistory,
                .createPost,
                .getFeedAsync,
                .likePost,
                .unlikePost,
                .getStoryFeed,
                .createStory,
                .getComments,
                .addComment,
                .watchedStory,
                .hidePost,
                .deletePost,
                .reportPost,
                .deleteStory,
                .getUserPostsAsync,
                .confirmationFollow,
                .getBlockUser,
                .addBlockUser,
                .removeBlockUser,
                .search:
            return ApplicationContext.shared.authToken
        default:
            return nil
        }
    }
}

File: core/api/src/main/java/com/iamkurtgoz/core/api/service/social/SocialService.kt
/*
 * Copyright 2024 Sporthor Android
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.iamkurtgoz.core.api.service.social

import androidx.annotation.Keep
import com.iamkurtgoz.core.network.model.BaseResponse
import com.iamkurtgoz.data.model.BlockedUsersResponseModel
import com.iamkurtgoz.data.model.CreatePostResponseModel
import com.iamkurtgoz.data.model.FollowUserResponseModel
import com.iamkurtgoz.data.model.GetCommentsResponseModel
import com.iamkurtgoz.data.model.GetFeedAsyncResponseModel
import com.iamkurtgoz.data.model.GetStoryFeedResponseModel
import com.iamkurtgoz.data.model.LikePostResponseModel
import com.iamkurtgoz.data.model.SearchHistorySocialResponseModel
import com.iamkurtgoz.data.model.SearchSocialResponseModel
import com.iamkurtgoz.data.model.UserRelationResponseModel
import com.iamkurtgoz.domain.model.request.AddCommentRequest
import com.iamkurtgoz.domain.model.request.AddSearchHistoryRequest
import com.iamkurtgoz.domain.model.request.CreatePostRequest
import com.iamkurtgoz.domain.model.request.CreateStoryRequest
import com.iamkurtgoz.domain.model.request.DeletePostRequest
import com.iamkurtgoz.domain.model.request.DeleteStoryRequest
import com.iamkurtgoz.domain.model.request.FollowUserRequest
import com.iamkurtgoz.domain.model.request.HidePostRequest
import com.iamkurtgoz.domain.model.request.LikePostRequest
import com.iamkurtgoz.domain.model.request.RemoveSearchHistoryRequest
import com.iamkurtgoz.domain.model.request.ReportPostRequest
import com.iamkurtgoz.domain.model.request.WatchedStoryRequest
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

@Keep
interface SocialService {
    @GET("Social/Search")
    suspend fun getSearch(@Query("SearchTerm") searchTerm: String, @Query("Role") role: Int? = null): Response<BaseResponse<SearchSocialResponseModel>>

    @GET("Social/GetSearchHistory")
    suspend fun getSearchHistory(): Response<BaseResponse<SearchHistorySocialResponseModel>>

    @POST("Social/AddSearchHistory")
    suspend fun addSearchHistory(@Body body: AddSearchHistoryRequest): Response<BaseResponse<Unit>>

    @POST("Social/RemoveSearchHistory")
    suspend fun removeSearchHistory(@Body body: RemoveSearchHistoryRequest): Response<BaseResponse<Unit>>

    @GET("Social/GetFollowers")
    suspend fun getFollowers(@Query("UserId") userId: String?): Response<BaseResponse<UserRelationResponseModel>>

    @GET("Social/GetFollowing")
    suspend fun getFollowing(
        @Query("UserId") userId: String?,
        @Query("role") role: Int? = null,
    ): Response<BaseResponse<UserRelationResponseModel>>


    @POST("Social/FollowUser")
    suspend fun followUser(@Body body: FollowUserRequest): Response<BaseResponse<FollowUserResponseModel>>

    @POST("Social/UnFollowUser")
    suspend fun unFollowUser(@Body body: FollowUserRequest): Response<BaseResponse<FollowUserResponseModel>>

    @GET("Social/GetFeedAsync")
    suspend fun getFeedAsync(@Query("Page") page: Int?, @Query("PageSize") pageSize: Int?): Response<BaseResponse<GetFeedAsyncResponseModel>>

    @GET("Social/GetStoryFeed")
    suspend fun getStoryFeed(): Response<BaseResponse<GetStoryFeedResponseModel>>

    @GET("Social/GetComments")
    suspend fun getComments(@Query("postId") postId: String?): Response<BaseResponse<GetCommentsResponseModel>>

    @POST("Social/CreatePost")
    suspend fun createPost(@Body body: CreatePostRequest): Response<BaseResponse<CreatePostResponseModel>>

    @POST("Social/DeletePost")
    suspend fun deletePost(@Body body: DeletePostRequest): Response<BaseResponse<Unit>>

    @POST("Social/ReportPost")
    suspend fun reportPost(@Body body: ReportPostRequest): Response<BaseResponse<Unit>>

    @POST("Social/HidePost")
    suspend fun hidePost(@Body body: HidePostRequest): Response<BaseResponse<Unit>>

    @POST("Social/LikePost")
    suspend fun likePost(@Body body: LikePostRequest): Response<BaseResponse<LikePostResponseModel>>

    @POST("Social/UnlikePost")
    suspend fun unlikePost(@Body body: LikePostRequest): Response<BaseResponse<LikePostResponseModel>>

    @POST("Social/AddComment")
    suspend fun addComment(@Body body: AddCommentRequest): Response<BaseResponse<Unit>>

    @POST("Social/DeleteStory")
    suspend fun deleteStory(@Body body: DeleteStoryRequest): Response<BaseResponse<Unit>>

    @POST("Social/CreateStory")
    suspend fun createStory(@Body body: CreateStoryRequest): Response<BaseResponse<Unit>>

    @POST("Social/WatchedStory")
    suspend fun watchedStory(@Body body: WatchedStoryRequest): Response<BaseResponse<Unit>>

    @GET("Social/GetBlockUser")
    suspend fun getBlockedUsers(): Response<BaseResponse<BlockedUsersResponseModel>>
}
